xxx-package-module-property.md

```diff
@@ -0,0 +1,273 @@
+behaves in the edge cases and for submodule requires (`pkg/x` imports). Since
+tools are currently driving the ecosystem conventions, it is worth refining the
+exact conventions with an active specification that can gain support, so that
+we can continue to converge on the module contract in NodeJS, and do our best
+to avoid incompatibilities in future.
+
+## 4. Proposal
+
+Instead of trying to consider a single unified resolver, we break the behaviour
+of NodeJS resolution into two separate resolvers:
+* The current resolver as in use today, which will continue to be used to
+resolve CommonJS modules from CommonJS modules, to ensure absolutely no
+breaking edge cases.
+* The new ES Modules resolver, that also has the ability to load CommonJS
+modules based on a small extension to the existing resolution algorithm.
+
+When using CommonJS `require`, the legacy resolver would be applied, and when
+using ES modules, the new ES module resolver algorithm, as along the lines
+specified here would be applied.
+
+**The rule proposed here is that the ES module resolver always loads a module
+from package with a "module" property as an ES module, and loads a module from
+a package without that property as a CommonJS module (unless it is a .mjs file
+or "use module" source).**
+
+Under this rule, the simple cases remain the same as the DDJS proposal:
+
+* A package with only a `main` and no `module` property will be loaded as
+containing CommonJS modules only.
+* A package with only a `module` property and no `main` property will be loaded
+as containing ES Modules only.
+
+The difficult case with the DDJS proposal is the transition case of a package
+that contains both a `main` and `module` property - selecting which main entry
+point and target to use when loading `pkg` or `pkg/x.js`.
+
+For a package that contains both a `main` and a `module` property -
+* When the parent module doing the require is an ES Module, the `module` main
+will apply, and any module loaded from the package will be loaded as an ES Module.
+* When the parent module doing the require is a CommonJS module, the `main`
+main will apply, and any module loaded from the package will be loaded as
+a CommonJS Module.
+
+In this way, we continue to support the existing ecosystem with backwards
+compatibility, while keeping the scope of the specification as simple as possible.
+
+## 4.1 Public API for Mixed CJS and ES Module Packages
+
+A package delivering both CommonJS and ES Modules would then typically
+tell its users to just import via `import {x} from 'pkgName'` or
+`require('pkgName').x`, with the `module` and `main` properties applying
+respectively.
+
+In the case where a package publicly exposes sub-modules, it would need
+to document that the CommonJS and ES Module sources are at different paths -
+`import {x} from 'pkgName/submodule.js'` and
+`import {x} from 'pkgName/cjs/submodule.js'`. Or simply a `.js` and
```

Comment:
I would love to see module.root to be added as well. This would allow hiding implementation details for modules and enable painless migartion/deprecation.